### ISSUE TITLE
What is the scope of the proof-of-concept wallet?

### ISSUE THREAD

**Issue by Arxcis** [din-ssi/wallet#7]
Ref bachelor #23: Identity wallet
>As part of the bachelor project, the team will develop a proof-of-concept reference implementation of an **open digital wallet (in form of smartphone app), that combines the notion of digital identity and user-centric, privacy preserving data storage.**

What is the intended scope of the wallet?
A wallet could be single-purpose, or it could be an "operating-system" for the decentralized web, or somewhere in between. The range is **HUGE**.
Should the scope of the wallet drive the scope of the SDK? Or should the SDK's scope drive the scope of the wallet? Or should they be independentently scoped?

**Comment by gtl-hig** [comment]
Both are very good questions. It is a bit chicken-and-egg situation, and, the process needs to be iterative - waterfall will not work here. Also, scoping will need to follow agile methodology, because we are working with new technology, rather than "work as usual". So, I think, those two are given:
* agile methodology
* iterative process

As for what to start with, I tend to lean more towards starting with a simple wallet for single purpose, and then, adding more to it. Starting with something tangible and simple, and expanding the scope, at each iteration. The advantage of this is that we can have something tangible early. The disadvantage is that certain things will have to be refactored and re-done, because the choices for "that single purpose" may not be adequate when we expand the scope.  So, refactoring and perhaps even "redoing" some elements will be necessary.   But, I feel there is no avoiding it, because this is not "work as usual" and none of us is an expert that could lay out a good abstract interoperable architecture that when instantiated would just nicely work.  The "discovery" of what is the best architecture, or the best SDK, needs to be "discovered" through the iterative process and re-factoring. 

So I suggest, starting with simple scope for single-purpose wallet. The adding additional use case. Then sketching initial SDK that could have, retrospectively, helped in actually DOING this wallet, if we were to do it again. And then, gradually, expanding the scope.

**Comment by Arxcis** [comment]
Ok. Maybe we need to ask a different question like, what is the minimum wallet?

From my understanding, agile is about keeping the scope flexible but minimal at any given point. They go hand in hand. The scope is allowed to be minimal, because the scope is flexible. 

Waterfall forces the opposite behaviour, because you have to define the scope at the beginning, and it may lead to scope maximization.

**Comment by Arxcis** [comment]
> The disadvantage is that certain things will have to be refactored and re-done, because the choices for "that single purpose" may not be adequate when we expand the scope. So, refactoring and perhaps even "redoing" some elements will be necessary.

I don't see this is as much as an disadvantage, and more like an unavoidable feature of software development. We might as well embrace it fully, than try to pretend it wont happen. I think this happens in both waterfall and agile, but the difference being that in agile you embrace the inevitable change, but in watefall you try to avoid it.

**Comment by Arxcis** [comment]
> So I suggest, starting with simple scope for single-purpose wallet.

Ok. So we agree to do a single-purpose wallet. Now we need to figure out what the single-purpose the wallet should serve. I think this can be discussed in a separate thread. Of course the different purposes require different scopes, so still keep that in mind.

**Conclusion of the scoping process**
1. Implement a wallet serving a single purpose.
2. Design the SDK retroactively, to meet the requirements of the single-purpose wallet.
3. Implement another wallet, or extend existing wallet, serving a new single purpose.
4. Repeat 2 and 3.

_The process expressed as c-code_
```c
do {
  Wallet wallet = makeWallet();
  makeSureSDKMeetsWalletRequirements(&wallet);
} while(notTired)
```

**Comment by vongohren** [comment]
@Arxcis @gtl-hig i agree on most of this, and yes we need to scope things small as we go, that also means more work as we need to do rescoping from time to time as we hit our features.

A couple of immediate questions.
When is the start time of the project?
Are there other interested?

Lets keep it small, lets keep it minimal.
But lets proove interopability between certain wallets.
This is still scoping that needs to be done. We are collecting the thoughts as we go, but was just not expecting such eagerness before it all started :D So patience is needed Obi-Wan Kenobi.
Some thoughts from the community: https://drive.google.com/drive/folders/1zUn3uK5uODe0_RC_vum_lu_0iGZrKC6n?usp=sharing

**Comment by Arxcis** [comment]
:pray:

**Comment by vongohren** [comment]
@Arxcis did you have a reply to the questions above :D ?

**Comment by Arxcis** [comment]
Ah of course. I got totally immersed in the spirit of patience and Obi-wan Kenobi, so I totally forgot you had questions :sweat_smile: 

> When is the start time of the project?

Bachelor projects are scheduled to start from **01. january 2021**. But school recommends to get familiar with tools and frameworks before that, which is why I am here already.

> Are there other interested?

AFAIK none has shown any interest besides me (wierd right!?). I have been contacted by other groups, from other projects, looking for a 3rd or a 4th person. We will know for sure the **18th november**, how many are interested. I heard 4 groups were competing for a single project, so once that is assigned to 1 of them, there will be 3 groups roaming around LFG again. I am positive the coming days will bring in some more people.

If not, I am not sure what to do.

**Comment by vongohren** [comment]
Sweet, thanks for great info!

Yeah I do still believe you will write a great thesis alone, and this subject is so important that it will be a game changer for the path you choose onwards :P The community is also super helpful so there will be a lot of resources available for you. We just need to be mindfull of making the scope fitting!

I wish I had the possibility to do something cool like this, globally changing subject, before I signed my ass of to a consultancy company while I was at school :P Consultancy companies do kill the spirit of making ground breaking tech

**Comment by gtl-hig** [comment]
I will update you guys later in a week. We will have internal check on the groups/decisions about who is assigned what tasks, and I think that may take some time.

**Comment by Arxcis** [comment]
> groups/decisions about who is assigned what tasks,

I am waiting for this as well.

**Comment by vongohren** [comment]
@Arxcis @gtl-hig I was contacted by another group of 4 people. Who seemed interessted.
I dont know if I encouraged them or scared them away. But I guess we will wait and see :D 

They knew you @Arxcis  :D

**Comment by gtl-hig** [comment]
I was told not to say anything official. But, I was allowed to tell @Arxcis not to stop working. I believe (unofficially) the project will go ahead as a individual BSc. We need to wait for the official announcement. In the meantime:

* @Arxcis - see issue #3 such that we can wrap up the decision if we stay here, or move the discussions "there".
* everyone - think WHO is the primary audience for this project. If the primary audience are developers, and the proof-of-concept wallet is only done as "reference implementation" that demonstrates the capabilities of the SDK, then this issue needs to be postponed, and we need to focus on #8 first.

**Comment by Arxcis** [comment]
>  But, I was allowed to tell @Arxcis not to stop working. I believe (unofficially) the project will go ahead as a individual BSc.

Great news - that the project is allowed to move forward :+1: Too bad none else showed up :-1: Nevertheless, I am confident that we will be able to produce something valuable, even with an individual BSc.

**Comment by Arxcis** [comment]
>     * everyone - think WHO is the primary audience for this project.

I got very confused about `WHO`  I thought you meant World Health Organization https://www.who.int/, but I now realize you meant **who**? :sweat_smile:

Ok so who needs an SDK for Android? Is it an open problem for SSI developers? Is there a lack of a shared codebase for SSI-related stuff on Android? 

On the flip side. Is the current SSI-wallets on Android lacking? As you said @gtl-hig , who should benefit from this project? Developers on Android, or end-users on Android? Who should we focus on? What and who's problem are we trying to solve specifically?

**Comment by vongohren** [comment]
Well me as a company, Diwala, is still in need of and open source SDK for mobile wallets. Or open source SDK for cloud agents. There exists the aeries agent, but it is still locked to hyperledger, sovrin, which is unfortunate. There exists some xamarin frameworks for mobile, but again sovrin hyperledger. It also exists go library that they say is completly fine to compile to android and iOS, and have examples on it. But again, aeries with sovrin and hyperledger.

You ahve the JS lib of uPort, DAF-framework, which might be feasable, but that only works for itself. So does aeries in that manner. With work for itself i mean that you cannot take DAF and aeries and have a interopexchange between two wallets that just work.

So the space is very fragemented in one way, as the hyperledger aeries has gotten a whole lot longer than the other guys, but the support for any og aeries sucks.
